### ISSUE TITLE
Anchor build mutable accounts check

### ISSUE THREAD

**Issue by fedorerd** [coral-xyz/anchor#2297]
While doing some stuff, just mentioned one thing (which is probably anchor related). There is no warnings/errors popping during anchor build if you:
1. Add account to the context without (mut) mark.
2. Borrow account as mutable with `&mut` inside instruction.
3. Modify any value.

There were no warning and build completed successfully. Moreover, I was able to mutate borrowed value inside instruction and work with updated value, but changes didn't reflect on solana anyhow.

More specifically, what I had is:
Context:
```
#[derive(Accounts)]
pub struct Ix<'info> {
    #[account(mut)]
    pub a: Account<'info, Raffle>,
    pub b: Account<'info, Rids>
}
```
Instruction:
```
pub fn ix(ctx: Context<Ix>) -> Result<()> {
    let a = &mut ctx.accounts.a;
    let b = &mut ctx.accounts.b;
    
    b.field += 1;
    a.field = b.field;

    Ok(())
}
```

Build was successfull. During test, `b.field` was equal to `0`. After invoking this ix, `a.field` was set to `1` (`b.field + 1`), and `b.field` didn't change.

So, basically, I believe it would be cool if such stuff could be checked by anchor during build.

**Comment by Henry-E** [comment]
It's been a known issue for a while that if you don't mark an account as `mut` and still try to use it as `mut` in the account struct anchor will not complain. 

If you have any suggestions for an implementation to check whether an account that's being used a `mut` has been marked as `mut` in the derive struct, do let us know.

**Comment by munanadi** [comment]
Can this be caught with a linting extension? Or do we need an LSP for this? 
I believe a linting extension will be able to detect this.
If yes. What are the other types of things that can be checked? 
Can we open an issue and make a list of such lint errors?

**Comment by chen-robert** [comment]
Resolved in v2, will not land in v1 because it's a breaking change
